docs/markdown/Writing Plugins/tutorials/02-integrate-goal-with-the-repository.md

```diff
@@ -0,0 +1,663 @@
+    ]
+```
+
+In this file, we use an advanced feature of Pants, [union rules](https://www.pantsbuild.org/docs/rules-api-unions):
+
+```python
+def rules():
+    return [
+        *collect_rules(),
+        UnionRule(PutativeTargetsRequest, PutativeProjectVersionTargetsRequest),
+    ]
+```
+
+When the `tailor` goal is run, the build graph is analyzed to see when `PutativeTargetsRequest` is needed, i.e. to find out if there are any files (yet unknown to Pants) that look like they could potentially be made targets. For instance, if there is a `requirements.txt` file, a `python_requirement` target is created and when there is a Python `test_` module, a `python_test` target is created. To be able to customize the `tailor` goal (to allow generation of custom targets), we need to "extend" the build graph. That is, we ask Pants to also run our rule when searching for files that maybe should have a target created.
+
+We also have to make sure that the new rule is collected:
+
+```python pants-plugins/project_version/register.py
+...
+def rules():
+    return [*project_version_rules.rules(), *tailor_rules.rules()]
+```
+
+Let's remove existing `version_file` target from the `myapp/BUILD` file and run the `tailor` goal:
+
+```
+$ ./pants tailor ::        
+Created myapp/BUILD:
+  - Add version_file target project-version-file 
+```
+
+If you have multiple projects, being able to generate the targets automatically may save time. You would also likely want to run the `tailor` goal in the check mode to confirm that new projects created have a `version_file` target. Remove the `version_file` target from the `myapp/BUILD` file and re-run the `tailor` goal:
+
+```
+$ ./pants tailor --check ::                         
+Would create myapp/BUILD:
+  - Add version_file target project-version-file
+
+To fix `tailor` failures, run `./pants tailor`.
+```
+
+## Running system tools
+
+Pants lets you [run system applications](https://www.pantsbuild.org/docs/rules-api-installing-tools) your plugin may need. For our use case, we can assume that Git is installed and can be run from the `/usr/bin/git`. If there's a `VERSION` file in the root of the repository representing the final artifact version (in case of a monolith), we could use Git to confirm that the version string matches the latest tag the repository was tagged with.
+
+We can create a new rule:
+
+```python
+class GitTagVersion(str):
+    pass
+
+@rule
+async def get_git_repo_version() -> GitTagVersion:
+    git_paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(
+            binary_name="git",
+            search_path=["/usr/bin", "/bin"],
+        ),
+    )
+    git_bin = git_paths.first_path
+    if git_bin is None:
+        raise OSError("Could not find 'git'.")
+    git_describe = await Get(
+        ProcessResult,
+        Process(
+            argv=[git_bin.path, "-C", BuildRoot().path, "describe", "--tags"],
```

Comment:
Would recommend to take `BuildRoot` as an input arg, like: 

https://github.com/pantsbuild/pants/blob/06a8d517dba5ce1502917390f3d3497d28400967/src/python/pants/core/util_rules/distdir.py#L30

then use that here. ;)

Reply:
Done!